File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.cc
#include <iostream>
#include <queue>
using namespace std;

int T;
int L;
int startX, startY;
int targetX, targetY;
int dirX[] = {2, -2, 2, -2, 1, -1, 1, -1};
int dirY[] = {1, 1, -1, -1, 2, 2, -2, -2};

int graph[301][301];
queue<pair<int, int>> q;

void Reset()
{
    while (!q.empty())
        q.pop();

    for (int i = 0; i < 301; i++)
    {
        for (int j = 0; j < 301; j++)
        {
            graph[i][j] = 0;            
        }
    }
}

void BFS(int x, int y)
{
    q.push({x,y});
    while (!q.empty())
    {
        int a = q.front().first;
        int b = q.front().second;
        q.pop();

        if (a == targetX && b == targetY)
        {
            cout << graph[a][b] << endl;
            Reset();
            break;
        }

        for (int i = 0; i < 8; i++)
        {
            int na = a + dirX[i];
            int nb = b + dirY[i];

            if (na < 0 || na >= L || nb < 0 || nb >= L) continue;
            if (graph[na][nb] != 0) continue;
            
            q.push({na, nb});
            graph[na][nb] = graph[a][b] + 1;
        }
    }
}

int main()
{
    cin >> T;
    
    for (int i = 0; i < T; i++)
    {
        cin >> L;
        cin >> startX >> startY;
        cin >> targetX >> targetY;

        BFS(startX, startY);
    }
    
    return 0;
}

File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.cs
using System;
using System.Collections.Generic;

namespace MyCompiler {
    class Program {
        static int[] dirX = {2, -2, 2, -2, 1, -1, 1, -1};
        static int[] dirY = {1, 1, -1, -1, 2, 2, -2, -2};
        static int L;
        static int targetX, targetY;
        static int [,] graph;

        static void BFS(int x, int y)
        {
            Queue<Tuple<int, int>> queue = new Queue<Tuple<int, int>>();
            queue.Enqueue(new Tuple<int, int>(x, y));

            while (queue.Count > 0)
            {
                Tuple<int, int> curr = queue.Dequeue();

                if (curr.Item1 == targetX && curr.Item2 == targetY)
                {
                    Console.WriteLine(graph[targetX, targetY]);
                    break;
                }

                for (int i = 0; i < 8; i++)
                {
                    int newX = curr.Item1 + dirX[i];
                    int newY = curr.Item2 + dirY[i];

                    if (newX < 0 || newX >= L || newY < 0 | newY >= L) continue;
                    if (graph[newX, newY] != 0) continue;

                    graph[newX, newY] = graph[curr.Item1, curr.Item2] + 1;
                    queue.Enqueue(new Tuple<int, int>(newX, newY));
                }
            }
        }
        
        public static void Main() {
            int N = int.Parse(Console.ReadLine());

            for (int i = 0; i < N; i++)
            {
                L = int.Parse(Console.ReadLine());
                string[] start = Console.ReadLine().Split();
                string[] target = Console.ReadLine().Split();
                graph = new int[311, 311];

                targetX = int.Parse(target[0]);
                targetY = int.Parse(target[1]);

                BFS(int.Parse(start[0]), int.Parse(start[1]));
            }
            
        }
    }
}